pkg/storage/allocator_scorer.go

```diff
@@ -552,13 +552,30 @@ func shouldRebalance(
 	store roachpb.StoreDescriptor,
 	sl StoreList,
 	rangeInfo RangeInfo,
+	existingNodeLocalities map[roachpb.NodeID]roachpb.Locality,
 ) bool {
+	// Rebalance if this store is too full.
 	if store.Capacity.FractionUsed() >= maxFractionUsedThreshold {
 		log.VEventf(ctx, 2, "s%d: should-rebalance(disk-full): fraction-used=%.2f, capacity=(%v)",
 			store.StoreID, store.Capacity.FractionUsed(), store.Capacity)
 		return true
 	}
 
+	// Rebalance if we could achieve better locality diversity by doing so.
+	diversityScore := diversityRemovalScore(store.Node.NodeID, existingNodeLocalities)
+	for _, desc := range sl.stores {
+		if !preexistingReplicaCheck(desc.Node.NodeID, rangeInfo.Desc.Replicas) {
+			continue
+		}
+		otherScore := rebalanceToDiversityScore(desc, existingNodeLocalities)
+		if otherScore > diversityScore {
```

Comment:
@a-robinson , This is great!

I have some confusion here, how could we gurantee we could achieve better locality for this range if `otherScore > diversityScore`?